rfcs/converged-implementation-patterns.md

```diff
@@ -0,0 +1,258 @@
+import { SampleState } from './Sample.types';
+
+/**
+* Styles for the root slot
+*/
+export const useRootStyles = makeStyles([
+  [
+    null,
+    theme => ({
+      backgroundColor: theme.neutralColorTokens.neutralBackground1,
+      paddingRight: '12px',
+    }),
+  ],
+]);
+
+/**
+* Styles for the icon slot, uses state selectors from SampleState
+*/
+export const useIconStyles = makeStyles<Pick<SampleState, 'someState'>>([
+  [
+    // Conditionally apply styles
+    (someState) => someState == 1 ? true : false,
+    () => ({
+      width: '20px',
+      height: '20px',
+      marginRight: '9px',
+    }),
+  ],
+]);
+
+/**
+* Applies style classnames to slots
+*/
+export const useSampleStyles = (state: SampleState) => {
+  const rootClassName = useRootStyles({});
+  const iconClassName = useIconStyles({ someState: state.someState });
+
+  // ax is a util that deduplicates classnames
+  state.className = ax(rootClassName, state.className);
+
+  if (state.icon) {
+    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+    // @ts-ignore
+    // currently shorthand slot typings are broken
+    state.icon.className = ax(iconClassName, state.icon.className);
+  }
+};
+```
+
+#### renderSample.tsx
+
+Renders the correct JSX output of the component and its slots given the correct state.
+
+This should be a pure function whose sole responsibility is to render JSX from the provided state. No state mutation or processing
+should happen in this function, but rather done in the `useSample` hook
+
+```typescript
+import { getSlots } from '@fluentui/react-utils';
+import { sampleShorthandProps } from './useSample';
+
+// state should come from `useSample` hook
+export const renderSample = (state: SampleState) => {
+  const { slots, slotProps } = getSlots(state, sampleShorthandProps);
+
+  return (
+    <slots.root {...slotProps.root}>
+      <slot.icon {...slotProps.icon} />
+    </slots.root>
+  );
+};
+```
+
+The `state` stores all information on the slots/tags/JSX that should be rendered after processing by the relevant hook. The `getSlots` utility extracts known slot information such as tag names and props.
+
+See below for how `state` might contain useful props for rendering.
+
+#### Sample.types.ts
+The below will probably be present in every component, other utility types are fair game if they are necessary.
+
+```typescript
+import * as React from 'react';
+import { ComponentProps, ShorthandProps } from '@fluentui/react-utils';
+
+// For a component all HTML attributes should be allowed to maximize consistencty with DOM
+export interface SampleProps extends ComponentProps, React.HTMLAttributes<HTMLElement> {
+  /** Icon slot as a shorthand */
+  icon?: ShorthandProps;
+}
+
+// Component state extends props and also adds other internal state used in the lifecycle of the component
+// For example for a popup component tracking an `open` state as boolean
+export interface MenuItemState extends SampleProps {
+  open: boolean;
```

Comment:
```suggestion
  icon?: ObjectSlotProp

  open: boolean;
```

All slots will be resolved to object in `*State`

Reply:
Agreed

Reply:
One more there, `ref` should be also typed

Reply:
Updated with a describing comment